working/1426-extension-types/feature-specification-implements-rep.md

```diff
@@ -0,0 +1,227 @@
+  A(this.i);
+}
+
+var a = A(2);
+int i = a; // OK.
+List<int> list = <A>[a]; // OK.
+```
+
+This subtype relationship is sound because the run-time value of an
+expression of type `A` is an instance of `int`. 
+
+This subtype relationship may be desirable in the case where there is no
+need to protect an object accessed as an `A` against being accessed as an
+`int`, and it is even considered to be useful that it will readily "become
+an `int` whenever needed".
+
+We need to introduce the _ultimate representation type_ of an inline type
+`V<T1 .. Tk>`: This is the instantiated representation type `R` of `V` if
+that is a type that is not and does not contain any inline
+types. Otherwise, assume that `R` is `V1<U1 .. Us>`, then the ultimate
+representation type of `V` is the ultimate representation type of
+`V1<W1 .. Ws>` where `Wj` is the ultimate representation type of `Uj`, for
+`j` in `1 .. s`. Similarly for function types and other composite types.
+
+*In this document it is assumed that the ultimate representation type
+exists. This is ensured by means of a static check on each inline class
+declaration, as specified in the inline class feature specification.*
+
+The corresponding change to the feature specification is be that the
+following sentence in the section
+[Composing inline classes](https://github.com/dart-lang/language/blob/main/accepted/future-releases/inline-classes/feature-specification.md#composing-inline-classes)
+is adjusted:
+
+> A compile-time error occurs if _V1_ is a type name or a parameterized type which occurs as a superinterface in an inline class declaration _DV_, but _V1_ does not denote an inline type.
+
+It is adjusted to end as follows:
+
+> ... declaration _DV_, unless _V1_ denotes an inline type, or _V1_ is a supertype of the ultimate representation type of _DV_.
+
+
+Moreover, a compile-time error occurs if the implements clause of _DV_
+contains two or more types that are non-inline types. A compile-time error
+occurs if the implements clause of _DV_ contains a non-inline type that
+occurs in any other position than the last one.
+
+*This ensures that it is only necessary for a reader of the declaration to
+check the last implemented type in order to see whether or not it is using
+this feature. It is somewhat similar to the rule that `super(...)` can only
+occur as the last element in a constructor initializer list.*
+
+Let _DV_ be an inline class declaration named `V` with representation type
+`R` and assume that the `implements` clause of _DV_ includes the non-inline
+type `R0`. *We then have `R <: R0`, because anything else is an error.*
+
+Invocations of members declared by `V` or declared by an inline
+superinterface of _DV_ are unaffected by the fact that _DV_ implements
+`R0`.
+
+Let `m` be a member name which is not declared by _DV_. Assume that the
+interface of `R0` has a member named `m`. A compile-time error occurs if an
+inline superinterface of _DV_ also declares a member named `m`.
```

Comment:
Should we use _basename_ here, to avoid allowing an inline superinterface method and an non-inline class setter?

Reply:
Weird exception: We _do_ require getters and setters to be provided together as extension methods:

```dart
class A {
  set p(int i) {}
}

extension on A {
  int get p => 0;
}

void main() {
  A().p = A().p;
}
```

We get an error for `A().p` as an expression, because we can't find the getter: We look up the basename and find the setter as an instance member of `A`, then decide that we shouldn't even try extension methods, and declare that there is no getter.

This might be considered consistent because getting a member from an extension could be considered as being a completely different thing than getting it from the class of the receiver, so maybe it's fine that we insist on bundling getters and setters such that they must be provided by the class together, or by an extension together, if we somehow have both of them.

But the argument would then be that inline classes have a more symmetric relationship to each other: _Every_ inline class member is similar to an extension method, and the `implements` relationship ensures that we can always detect every conflict and require that it is resolved explicitly. I think this justifies the ability to put together a getter `p` and a setter `p=` from inline classes `I1` respectively `I2` by means of `inline class I implements I1, I2 {...}`. We can do the same thing with regular classes, and `I1`/`I2` are arguably in the "same domain".

So the remaining question is: Should we treat basename conflicts involving non-inline superinterfaces differently?

Reply:
We require getter and setter to be provided together everywhere but in instance member inheritance.

For example we also consider it a conflict if two different imports provide a getter and a setter with the same base name.

The question here then becomes whether this is like instance member inheritance or not.
And whether inline class superinterfaces and non-inline-class superinterfaces (or supertypes) are somehow *different*. 
It should probably just work, no matter where you get the member from.

Reply:
Sounds good. This is what we get by never saying 'basename', so we should be OK.